be/src/olap/memtable_flush_executor.cpp

```diff
@@ -0,0 +1,177 @@
+    int32_t flush_queue_idx = _get_queue_idx(path_hash); 
+    *flush_handler = new FlushHandler(flush_queue_idx, this);
+    return OLAP_SUCCESS;
+}
+
+void MemTableFlushExecutor::init(const std::vector<DataDir*>& data_dirs) {
+    int32_t data_dir_num = data_dirs.size();
+    _thread_num_per_store = std::max(1, config::flush_thread_num_per_store);
+    _num_threads = data_dir_num * _thread_num_per_store;
+
+    // create flush queues
+    for (int i = 0; i < _num_threads; ++i) {
+        BlockingQueue<MemTableFlushContext>* queue = new BlockingQueue<MemTableFlushContext>(10);
+        _flush_queues.push_back(queue);
+    }
+    // create thread pool
+    _flush_pool = new ThreadPool(_num_threads, 1);
+    for (int32_t i = 0; i < _num_threads; ++i) {
+       _flush_pool->offer(boost::bind<void>(&MemTableFlushExecutor::_flush_memtable, this, i));
+    }
+
+    // _path_map saves the path hash to current idx of flush queue.
+    // eg.
+    // there are 4 data stores, each store has 2 work thread.
+    // so there are 8(= 4 * 2) queues in _flush_queues.
+    // and the path hash of the 4 paths are mapped to idx 0, 2, 4, 6.
+    int32_t group = 0;
+    for (auto store : data_dirs) {
+        _path_map[store->path_hash()] = group;
+        group += _thread_num_per_store;
+    }
+}
+
+MemTableFlushExecutor::~MemTableFlushExecutor() {
+    // shutdown queues
+    for (auto queue : _flush_queues) {
+        queue->shutdown();
+    }
+
+    // shutdown thread pool
+    _flush_pool->shutdown();
+    _flush_pool->join();
+
+    // delete queue    
+    for (auto queue : _flush_queues) {
+        delete queue;
+    }
+    _flush_queues.clear();
+
+    delete _flush_pool;
+}
+
+int32_t MemTableFlushExecutor::_get_queue_idx(size_t path_hash) {
+    std::lock_guard<SpinLock> l(_lock);
+    int32_t cur_idx = _path_map[path_hash];
+    int32_t group = cur_idx / _thread_num_per_store;
+    int32_t next_idx = group * _thread_num_per_store + ((cur_idx + 1) % _thread_num_per_store);
+    DCHECK(next_idx < _num_threads);
+    _path_map[path_hash] = next_idx;
+    return cur_idx;
+}
+
+std::future<OLAPStatus> MemTableFlushExecutor::_push_memtable(int32_t queue_idx, MemTableFlushContext& ctx) {
+    ctx.flush_id = _id_generator.fetch_add(1);
+    std::promise<OLAPStatus> promise;
+    std::future<OLAPStatus> fu = promise.get_future();
+    {
+        std::lock_guard<SpinLock> l(_lock);
+        _flush_promises[ctx.flush_id] = std::move(promise);
+    }
+
+    _flush_queues[queue_idx]->blocking_put(ctx);
+    return fu;
+}
+
+void MemTableFlushExecutor::_flush_memtable(int32_t queue_idx) {
+    while(true) {
+        MemTableFlushContext ctx;
+        if (!_flush_queues[queue_idx]->blocking_get(&ctx)) {
+            // queue is empty and shutdown, end of thread
+            return;
+        }
+
+        // if last flush of this tablet already failed, just skip
+        if (ctx.flush_handler->last_flush_status() != OLAP_SUCCESS) {
```

Comment:
If you call last_flush_status() here, this means that Executor know details of flush handler.
there are other options to do this
1. Executor just do and callback, and Flusher will add more task in callback function
2. make a context an interface who has an abstract function get_state() to tell if it should be scheduled.